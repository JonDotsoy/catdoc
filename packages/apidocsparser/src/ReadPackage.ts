import fs from "fs"
import yaml from "yaml"
import once from "lodash/once"
import { ReadToc } from "./lib/iterUris"
import path from "path"
import { TOCFormats } from "./dto/TOCFormats"
import * as TocPlus from "./toc-plus"
import * as Item from "./toc-plus/Item"
import util from "util"

export class ReadPackage {
  pathLike: string
  toc: ReadToc | undefined

  private constructor(pathLike: string) {
    this.pathLike = path.resolve(pathLike)
    const isFolder =
      fs.existsSync(pathLike) && fs.statSync(pathLike).isDirectory()

    if (!isFolder) {
      const err = new TypeError(`Path ${pathLike} is not a directory`)
      Error.captureStackTrace(err, ReadPackage.readPackage)
      throw err
    }
  }

  async prepareLoadToc() {
    const tocFileReaded = this.readTocFile()

    if (!tocFileReaded) return null

    return {
      toc: await ReadToc.loadToc(tocFileReaded.body, {
        baseDirectory: this.pathLike,
      }).prepare(),
    }
  }

  prepare = once(async () => {
    const tocLoaded = await this.prepareLoadToc()

    this.toc = tocLoaded?.toc

    return this
  })

  static readPackage = (k: string) => new ReadPackage(k)

  private readTocFile() {
    const tocFileDetected = this.detectTocFile()

    if (tocFileDetected?.format === TOCFormats.JSON)
      return {
        ...tocFileDetected,
        body: JSON.parse(fs.readFileSync(tocFileDetected.path, "utf-8")),
      }
    if (tocFileDetected?.format === TOCFormats.YAML)
      return {
        ...tocFileDetected,
        body: yaml.parse(fs.readFileSync(tocFileDetected.path, "utf-8")),
      }

    return null
  }

  private detectTocFile() {
    const pathLikeTocJson = `${this.pathLike}/toc.json`
    const pathLikeTocYaml = `${this.pathLike}/toc.yaml`
    const pathLikeTocYml = `${this.pathLike}/toc.yml`

    if (fs.existsSync(pathLikeTocJson))
      return { format: TOCFormats.JSON, path: pathLikeTocJson }
    if (fs.existsSync(pathLikeTocYaml))
      return { format: TOCFormats.YAML, path: pathLikeTocYaml }
    if (fs.existsSync(pathLikeTocYml))
      return { format: TOCFormats.YAML, path: pathLikeTocYml }

    return null
  }

  findByKeyToc(keyToc: string) {
    return this.toc?.findItemByKeyToc(keyToc)
  }

  findItemByKeyToc(keyToc: string) {
    const item = this.toc?.findItemByKeyToc(keyToc)
    if (item instanceof Item.Item) return item
  }

  toJSON(): any {
    return new (class ReadPackage {})()
  }

  [util.inspect.custom]: util.CustomInspectFunction = (depth, options) => {
    return `ReadPackage ${util.formatWithOptions(options, {
      path: path.relative(process.cwd(), this.pathLike),
      toc: this.toc,
    })}`
  }
}
