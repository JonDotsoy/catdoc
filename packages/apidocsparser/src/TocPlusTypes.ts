import * as toc from "sculltoc/schema"
import fs from "fs/promises"
import mimeTypes from "mime-types"
import YAML from "yaml"
import $RefParser, { dereference } from "json-schema-ref-parser"
import path from "path"

export namespace TocPlus {
  export type Common<T> = {
    _item: T
    keyToc: string
  }

  export class Item implements Common<toc.Item> {
    ["content-type"]?: string;
    ["charset"]?: string
    bytes?: number
    #body?: Buffer | $RefParser.JSONSchema

    constructor(
      readonly keyToc: string,
      readonly _item: toc.Item,
      readonly uri: string
    ) {
      this["content-type"] = mimeTypes.lookup(uri) || undefined
      this["charset"] = this["content-type"]
        ? mimeTypes.charset(this["content-type"]) || undefined
        : undefined
    }

    async prepare() {
      await this.readBuffer()
      return this
    }

    get body() {
      return this.#body
    }
    bodyJson(
      replacer?: ((this: any, key: string, value: any) => any) | undefined,
      space?: string | number | undefined
    ) {
      return JSON.stringify(this.#body, replacer, space ?? 2)
    }
    bodyYaml(options?: YAML.Options | undefined) {
      return YAML.stringify(this.#body, options)
    }

    async readBuffer() {
      const buffer = await fs.readFile(this.uri)
      this.bytes = buffer.length

      this.#body = await this.resolveRefs(
        this.updateRefProp(this.parseBody(buffer))
      )
    }

    private updateRefProp(body: any) {
      const intent = (
        obj: any,
        prop?: string,
        descriptor?: TypedPropertyDescriptor<any> & PropertyDescriptor,
        updateValue?: (newVal: any) => void
      ) => {
        if (typeof obj === "object" && !(obj instanceof Buffer)) {
          const entries = Object.entries(Object.getOwnPropertyDescriptors(obj))

          entries.forEach(([key, descriptor]) => {
            intent(descriptor.value, key, descriptor, (newVal) => {
              obj[key] = newVal
            })
          })

          return
        }
        if (prop === "$ref" && typeof obj === "string" && updateValue) {
          const parsed = path.parse(obj)
          if (!parsed.dir.startsWith("#")) {
            updateValue(path.join(`${path.dirname(this.uri)}`, `${obj}`))
          }
        }
      }
      intent(body)
      return body
    }

    private parseBody(buffer: Buffer) {
      const getEncoding = () => {
        if (this.charset && Buffer.isEncoding(this.charset)) {
          return this.charset
        }
        return "utf8"
      }
      if (this["content-type"] === "application/json")
        return JSON.parse(buffer.toString(getEncoding()))
      if (this["content-type"] === "text/yaml")
        return YAML.parse(buffer.toString(getEncoding()))
      return buffer
    }

    private async resolveRefs(body: any) {
      if (body instanceof Buffer) return body

      const res = await dereference(body)

      return res
    }
  }

  export type Divider = Common<toc.Divider>
  export type Group = Common<toc.Group> & { itemsPlus: Element[] }
  export type Element = Group | Item | Divider
  export type Elements = Element[]
}
