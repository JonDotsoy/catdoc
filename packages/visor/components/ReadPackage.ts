import { ReadPackage } from "apidocsparser"
import getConfig from "next/config"

export const generalReadPackage = ReadPackage.readPackage(
  getConfig().serverRuntimeConfig.pathRepo
)
