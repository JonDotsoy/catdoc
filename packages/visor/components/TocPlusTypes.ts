import { Group as TocGroup, Item as TocItem, Divider as TocDivider } from 'sculltoc/schema';


export type TocPlusCommon = { keyToc: string; };
export type TocPlusItem = TocItem & TocPlusCommon & { uriCode: string };
export type TocPlusDivider = TocDivider & TocPlusCommon;
export type TocPlusGroup = TocGroup & TocPlusCommon & { itemsPlus: TocPlus[]; };
export type TocPlus = TocPlusGroup | TocPlusItem | TocPlusDivider;
export type TocPlusItems = TocPlus[];

export type MapItems = {
    listItems: TocPlus[],
    mapKeys: { [k: string]: TocPlus }
    mapUris: { [k: string]: TocPlus }
    uris: string[],
    keysToc: string[],
}
