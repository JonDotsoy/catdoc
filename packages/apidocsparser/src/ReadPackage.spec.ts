import { ReadPackage } from "./ReadPackage"
import util from "util"
import { TocPlus } from "./TocPlusTypes"
import { writeFileSync } from "fs"

const pathDemoRepo = `${__dirname}/../../../demo`

describe("ReadPackage", () => {
  describe("", () => {
    it("should create a read package instance", async () => {
      await ReadPackage.readPackage(pathDemoRepo).prepare()
    })

    it("should return the item with the method `findItemByKeyToc`", async () => {
      const readPackage = await ReadPackage.readPackage(pathDemoRepo).prepare()

      const itemFound = readPackage.findItemByKeyToc("$.1.1")

      expect(itemFound).toBeInstanceOf(TocPlus.Item)
    })

    it("should return an undefined if not found item", async () => {
      const readPackage = await ReadPackage.readPackage(pathDemoRepo).prepare()

      const itemFound = readPackage.findItemByKeyToc("this not exists")

      expect(itemFound).toBeUndefined()
    })

    it("should print a detail read package", async () => {
      const readPackage = await ReadPackage.readPackage(pathDemoRepo).prepare()

      expect(
        util.formatWithOptions(
          { depth: Infinity, maxArrayLength: Infinity },
          readPackage
        )
      ).toMatchSnapshot()
    })

    it("should parse references", async () => {
      const readPackage = await ReadPackage.readPackage(pathDemoRepo).prepare()

      const item = readPackage.findItemByKeyToc("$.3")

      expect(item).toBeDefined()
      writeFileSync(`${__filename}.demo.yaml`, item!.bodyYaml())
      writeFileSync(`${__filename}.demo.json`, item!.bodyJson())
    })
  })
})
