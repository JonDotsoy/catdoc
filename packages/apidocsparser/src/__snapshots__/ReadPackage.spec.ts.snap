// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadPackage  should print a detail read package 1`] = `
"ReadPackage {
  path: '../../demo',
  toc: ReadToc {
    items: [
      Group {
        keyToc: '$.1',
        title: 'Documentation',
        items: [
          Documentation {
            keyToc: '$.1.1',
            uri: '../../demo/docs/howto.md',
            'content-type': 'text/markdown',
            charset: 'UTF-8',
            bytes: 38
          }
        ]
      },
      Group { keyToc: '$.2', title: 'APIS' },
      Reference {
        keyToc: '$.3',
        uri: '../../demo/reference/hello.yaml',
        'content-type': 'text/yaml',
        charset: 'UTF-8',
        bytes: 5719,
        openapiVersion: '3.1.0',
        info: { title: 'hello', version: '1.0' },
        paths: [
          {
            path: '/users/{userId}',
            method: 'get',
            title: 'Get User Info by User ID',
            description: 'Retrieve the information of the user with the matching user ID.'
          },
          {
            path: '/users/{userId}',
            method: 'patch',
            title: 'Update User Information',
            description: 'Update the information of an existing user.'
          },
          {
            path: '/user',
            method: 'post',
            title: 'Create New User',
            description: 'Create a new user.'
          },
          {
            path: '/car',
            method: 'post',
            title: 'Create Car',
            description: undefined
          }
        ],
        schemas: [ { title: 'User' } ]
      },
      Group {
        keyToc: '$.4',
        title: 'Pokedex',
        items: [
          Reference {
            keyToc: '$.4.1',
            uri: '../../demo/reference/pokedex.yaml',
            'content-type': 'text/yaml',
            charset: 'UTF-8',
            bytes: 1464,
            openapiVersion: '3.0.2',
            info: { title: 'API Pokemon', description: 'API Pokemon', version: '1.0' },
            paths: [
              {
                path: '/{playerId}/pokedex',
                method: 'put',
                title: 'update list pokedex',
                description: undefined
              },
              {
                path: '/{playerId}/pokedex',
                method: 'get',
                title: 'read schema',
                description: undefined
              }
            ],
            schemas: [ { title: 'PokemonInfo' }, { title: 'PokedexDescriptor' } ]
          }
        ]
      }
    ]
  }
}"
`;

exports[`ReadPackage  should print a detail read package 2`] = `ReadPackage {}`;

exports[`ReadPackage  should print a detail read package 3`] = `"{}"`;
